Support for IPv6 interfaces (#37)

bruinsg opened the pull request:
This merge request has the following changes:
1. Removed support for out of support frameworks
2. Support for IPv6 networks on Windows and Linux
3. Fix for unsupported addresses that would show up when remote devices restart

bruinsg committed:
Add support for net60

bruinsg committed:
Support interfaces that only have IPv6 (#37)

bruinsg committed:
Only support pure IPv6 interfaces on Linux

bruinsg committed:
Support for IPv6 and bug fixes for multi interface systems

When a remote device restarts or broadcasts it information
this is received on all interfaces. This means that interfaces
will have unreachable IP addresses in the service announcement.
By checking the originating IP address filtering is perfomed
and unreachable addresses are ignored.

bruinsg commented:
@tmds Any feed back or comments on this?

tmds commented:
Sorry, I had missed your PR.
I will make some time to look at this next week.

tmds commented on src/Tmds.MDns/MdnsSocket.cs:
```diff
@@ -0,0 +1,142 @@
+            socket.SetSocketOption(SocketOptionLevel.IP, SocketOptionName.MulticastInterface, IPAddress.HostToNetworkOrder(index));
+
+            socket.Bind(new IPEndPoint(IPAddress.Any, IPv4EndPoint.Port));
+            var tempEndpoint = new IPEndPoint(IPAddress.Any, IPv4EndPoint.Port);
```
remove `tempEndpoint`.

bruinsg replied:
Done

tmds commented on src/Tmds.MDns/MdnsSocket.cs:
```diff
@@ -0,0 +1,142 @@
+            }
+
+            socket.Bind(new IPEndPoint(IPAddress.IPv6Any, IPv6EndPoint.Port));
+            var tempEndpoint = new IPEndPoint(IPAddress.IPv6Any, IPv6EndPoint.Port);
```
remove `tempEndpoint`.

bruinsg replied:
Done

tmds commented on src/Tmds.MDns/MdnsSocket.cs:
```diff
@@ -0,0 +1,142 @@
+            var socket = new Socket(AddressFamily.InterNetworkV6, SocketType.Dgram, ProtocolType.Udp);
+            socket.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ReuseAddress, true);
+
+            // The MulticastInterface fails for IPv6, see: https://github.com/dotnet/runtime/issues/24255
```
This is an old issue about MulticastInterface not working for IPv6 on Linux.

By now, we should be able to:
```
socket.SetSocketOption(SocketOptionLevel.IPv6, SocketOptionName.MulticastInterface, interfaceIndex).
```
and this should work on Linux and Windows afaik.

bruinsg replied:
Yes, this works now. I did try this before but couldn't get it to work. Guess there were some other issues at that time. It works now on both Linux and Windows.

tmds commented on src/Tmds.MDns/MdnsSocket.cs:
```diff
@@ -0,0 +1,142 @@
+        public delegate void OnReceivedFrom(IPAddress from, MemoryStream stream);
+        private Socket _socket;
+        private IPEndPoint _endPoint;
+        private EndPoint _senderEndpoint;
```
Can you make `readonly` where possible.
And rename `_endPoint` to `_sendEndPoint`, and `_senderEndpoint` to `_receiveEndPoint`.

bruinsg replied:
Done

tmds commented on src/Tmds.MDns/MdnsSocket.cs:
```diff
@@ -0,0 +1,142 @@
+            IPAddress ip = IPv4EndPoint.Address;
+            socket.SetSocketOption(SocketOptionLevel.IP, SocketOptionName.AddMembership, new MulticastOption(ip, index));
+            socket.SetSocketOption(SocketOptionLevel.IP, SocketOptionName.MulticastTimeToLive, 1);
+            var info = new MdnsSocket(socket, IPv4EndPoint);
```
`return new MdnsSocket(socket, IPv4EndPoint);`

tmds commented on src/Tmds.MDns/MdnsSocket.cs:
```diff
@@ -0,0 +1,142 @@
+            IPAddress ip = IPv6EndPoint.Address;
+            socket.SetSocketOption(SocketOptionLevel.IPv6, SocketOptionName.AddMembership, new IPv6MulticastOption(ip, index));
+            socket.SetSocketOption(SocketOptionLevel.IPv6, SocketOptionName.MulticastTimeToLive, 1);
+            var info = new MdnsSocket(socket, IPv6EndPoint);
```
`return new MdnsSocket(socket, IPv6EndPoint);`

bruinsg replied:
Missed this one, will push it on my next change

tmds commented on src/Tmds.MDns/NetworkInterfaceHandler.cs:
```diff
@@ -32,15 +33,12 @@ public NetworkInterfaceHandler(ServiceBrowser serviceBrowser, NetworkInterface n
-            _receiveEventArgs.SetBuffer(_buffer, 0, _buffer.Length);
-            _receiveEventArgs.Completed += new EventHandler<SocketAsyncEventArgs>(OnReceive);
-#else
+            _socketInfo = new List<MdnsSocket>();
```
Rename `_socketInfo` to `_sockets`

bruinsg replied:
done

tmds commented on src/Tmds.MDns/NetworkInterfaceHandler.cs:
```diff
@@ -177,49 +181,60 @@ internal void OnServiceQuery(Name serviceName)
+                                   .FirstOrDefault();
+        }
+
+        private bool IsUnreachable(IPAddress receivedFrom)
```
Instead of this, can we use `SO_BINDTODEVICE`?

bruinsg replied:
I will have a look at this. Can this be used for both Linux and Windows using the SetRawSocketOption?

tmds replied:
Stepping back for a moment.

`IsUnreachable` filters the packets to the interface they are received on.
I would expect the socket to not receive packets for an interface it didn't request membership for.
So this is a false assumption?
Is it not true on Windows, on Linux, or both?

bruinsg replied:
It is false on Linux.
It works as long as the message was initiated from the service. So send a requests gets a reply on the same port.
But when a remote device reboots the message is received on all sockets. It will show up all on interfaces.

It is important for me as I want to only discover devices on a internal test network (IPv6 local link) and not the devices on the LAN.

bruinsg replied:
I have setup a new VM to test the behavior on Windows.
I made the IsUnreachable return false and on Windows the behavior is ok.
On Linux the ip addresses of the remote are received on both local adapters.
So, for example the IPv6%2 is not reachable as it is actually on %3. Also, as the addresses overwrite each other I can see
the ip addresses constantly change '~'

tmds commented on src/Tmds.MDns/LinuxHelper.cs:
```diff
@@ -0,0 +1,25 @@
+        public static unsafe void MultiCastV6(Socket socket, int index)
+        {
+#if NET6_0
+            int setVal = index;
```
See comment at the call site, this should no longer be needed.

bruinsg replied:
Done

tmds commented:
@bruinsg I took at the PR and added some comments.

I don't want to drop support for .NET framework, so we should add back a tfm for that.

Instead of having a `List<MdnsSocket>`, can we change `NetworkInterfaceHandler` to be `NetworkInterfaceHandlerSocket`, and take a `bool isIPv6` argument?
I think we'll reduce the lines of code changed, which will make it easier to review and avoid regressions.

bruinsg committed:
Addressing feed back on merge request

* Changed MdnsSocket to NetworkInterfaceHandlerSocket
* Use MulticastInterface that works for Linux and Windows

bruinsg committed:
Fix for IPv6 network not being ready when NetworkInterfaceHandler is created

During a system reboot the IPv6 is not always available when the NetworkHandler
is created. Listening to the NetworkAvailabilityChanged helps detecting changes
in the network.

bruinsg commented on src/Tmds.MDns/ServiceBrowser.cs:
```diff
@@ -234,9 +235,13 @@ private void CheckNetworkInterfaceStatuses(Dictionary<int, NetworkInterfaceHandl
                 {
+                    var hasIPv4 = networkInterface.Supports(NetworkInterfaceComponent.IPv4);
+                    // Pure IPv6 networks are only supported on Linux
+                    var hasIPv6 = networkInterface.Supports(NetworkInterfaceComponent.IPv6) && RuntimeInformation.IsOSPlatform(OSPlatform.Linux);
```
Fixed, it now works on Windows as well

bruinsg committed:
Support for netstandard 1.3

bruinsg commented:
I have address your concerns and added back support for netstandard1.3. Please have a look if there are any further issues you would like me address

tmds committed:
Refactor some things.

tmds commented:
Sorry this is moving slowly. I've been very busy the last weeks.

I took a close look at the PR now, and push a commit to your branch.
I've mostly refactored the code in a way the diff is easier to review and understand.
I also moved from netstandard1.3 to netstandard2.0 (to be able to simplify some code).

Can you take a look, and add any feedback you may have?

bruinsg committed:
Fixes on refactoring

StartReceive after receiving a message from non local addresses

bruinsg committed:
Fixing legacy code

Remove netstandard 1.3 and fix depreaction warnings

bruinsg commented:
Thanks for your response!

There was one bug in the code where the start receive was not issues after receiving a message from a non local ip address.
There were also some deprecation warnings that I fixed, please check if this is ok with you.

tmds commented:
> There was one bug in the code where the start receive was not issues after receiving a message from a non local ip address.

Thanks for the fix-ups.

> There were also some deprecation warnings that I fixed, please check if this is ok with you.

I like using `IPAddress.Address` for IPv4 (though it's obsolete) because it doesn't allocate.
Adding a pragma that ignores the obsolete is ok for me.
Leaving it as-is, is fine with me too.

Overall, this looks good to me.
I'm going to merge this on Friday and then make a new release of the library on nuget.org.

bruinsg commented:
Great, thanks. Looking forward to the new release

tmds commented:
This is part of v0.8.0 which has just been uploaded to nuget.org.

bruinsg commented:
Thanks.
